`ref` parameters, arguments, returns and `val` returns (#5434)
carbon-language/carbon-lang: proposals/p5434.md

    and
-   the left-hand operand of [assignment operators](/docs/design/assignment.md)
    such as `=` and `+=`.
-   [the `++` and `--` operators](/docs/design/assignment.md).

Even in these cases, the arguments will not be marked with `ref` at the call
site. (Generally the `ref` parameter is the `self` parameter, and so wouldn't be
marked. The exception is `BindToRef`, but we don't want to mark its argument
with `ref`.)

As an _experiment_, we are saying a pointer formed by taking the address of a
`ref` bound name is LLVM-`captures(none)` and LLVM-`noalias`. This means that
while a `ref` parameter could be passed into a function by address, the
restrictions also allow a "move-in-move-out" approach (once we define the move
operation), assuming it is not [marked `bound`](#bound-parameters). The intent
here is to leave the door open to a calling convention using registers and less
indirection for small-enough objects.

This means that the following code is invalid:

```carbon
fn F(ref a: i32, ref b: i32) -> bool;

fn G() -> bool {
  var v: i32 = 1;
  return F(ref v, ref v);
}
```

### `ref`, `let`, and `var` returns

> The return of a function can optionally be marked `ref` or `let`. These control

Reviewer: I'd like to suggest an alternative:
- The return of a function can optionally marked with `ref` or `val`, which causes the function call to be a reference or value expression (respectively).
- A binding pattern can optionally be marked with `ref` or `val`, which causes name expressions that name the binding to be reference or value expressions (respectively). Note that this gives us a way to declare a value binding that's nested inside a `var`.

That way there's a clear and consistent parallel between `ref` and `val`:
- Each can be used as a modifier on a binding pattern or function return declaration, and nowhere else.
- Each causes uses of the declared entity to have a particular category.
- Each is an abbreviation of the name of that category.
- One is the implicit default for bindings inside `var`, and the other is the implicit default for bindings outside `var`.

Even the ways the parallels break down are illuminating, because they reflect the asymmetries of the underlying model of expression categories:
- `ref` on a binding pattern constrains the category of the scrutinee, whereas `val` does not, but that reflects the fact that there are no conversions from other categories to durable reference.
- `var` is a modifier on `->` but acts as an independent operator rather than a binding modifier in patterns, but that reflects the fact that matching an initializing expression has side effects, and does not propagate its category to the subpattern's scrutinee.

This also preserves the current simplicity of `let`: it's solely a pattern introducer, with no connection to expression categories or mutability.

Reviewer 2: I forgot to update here but meant to -- I think roughly this idea is what I wrote up in the leads issue here: https://github.com/carbon-language/carbon-lang/issues/5522#issuecomment-2972029100

(I think I misunderstood it at first to be something different, but in a live discussion Geoff corrected me.)

Flagging that here to connect all the dots.

## Proposal

### `ref` bindings

We introduce a new keyword `ref`. This may be added to a `:` binding to mark it
as binding to a reference expression, as in:

```carbon
fn F(ptr: i32*) {
  // A reference binding `x`.
  let ref x: i32 = *ptr;

  // Use of `x` is a reference expression that
  // refers to the same object as `*ptr`.
  Assert(&x == ptr);

  // Equivalent to `*ptr += 1;`.
  x += 1;
}

fn G() {
  var y: i32 = 2;
  F(&y);
  Assert(y == 3);
}
```

The use of the name (`x` in the example) of a `ref` binding forms a durable
reference expression. We ensure that reference expressions formed by way of
reference bindings _do not dangle_. A `ref` binding may only bind to a durable
reference expression or an expression that can be converted to one. The bound
durable reference expression must outlive the `ref` binding.

The address of a `ref` bound name gives the address of the bound object, so
`&x == ptr` above. The reference itself does not have an address, and unlike a
pointer can't be rebound to reference a different object.

We remove `addr`, and use instead use `ref` for the `self` parameter when an
object is required. Note that the type will change from `Self*` to `Self` in

> this case.

Reviewer 2: Could you add a note here that we may re-add `addr` as a way to express uses of `self` that we want to disallow for `ref` bindings generally but need an escape hatch for migrated C++ code that relies on these patterns? (Happy for it to be a very brief note and a forward-reference to future work if you already cover this there.)

Author: Added future work.

    // Previously would dereference `self` in
    // the body of the method.
    self->x += 3;
  }

  // ✅ Now valid.
  fn NewMethod[ref self: Self]() {
    // Now `self` is a reference expression,
    // and is not dereferenced.
    self.x += 3;
  }

  // ✅ Other uses are unchanged.
  fn Get[self: Self]() -> i32 {
    return self.x;
  }

  var x: i32;
}
```

Potentially abbreviating the syntax further (to allow `ref self` as a short form
of `ref self: Self`) is left as future work.

The `ref` modifier is allowed on any `:` binding that's not inside a `var`
pattern.

```carbon
fn AddTwoToRef(ref x: i32) {
  x += 1;
  let ref y: i32 = x;
  y += 1;
}

// Equivalent to:
fn AddTwoToRef(ref x: i32) {
  x += 1;
  let y_ptr: i32* = &x;
  *y_ptr += 1;
}
```

We add support for `ref` and `var` in a
[struct pattern](/docs/design/pattern_matching.md#struct-patterns) when using
the shorthand `a: T` syntax for `.a = a: T`:

```carbon
let {var a: i32, ref b: i32} = ...;

// Now equivalent to:
let {.a = var a: i32, .b = ref b: i32} = ...;
```

The `ref` modifier is forbidden on the bindings in `class` or `struct` fields.

```
class Invalid {
  // ❌ Invalid.
  ref var m: i32;
}

> ```

Reviewer 2: The example says `ref var m: i32`, but that shouldn't be valid because `ref` has to be attached to the binding `m` and so has to come after the `var` introducer.

And you already said that `ref` isn't allowed inside of a `var`, so that makes just reversing the example seem a bit redundant.

Maybe something like:

````suggestion
The `ref` modifier is forbidden on the bindings in `class` or `struct` fields.

```
let OuterSize:! i32 = 123;

class Invalid {
  // ❌ Invalid.
  let ref Size:! i32 = OuterSize;
}
```
````

In a function argument list, arguments to non-`self` `ref` parameters are also
marked with `ref`. Continuing the example:

```carbon
var z: i32 = 3;
AddTwoToRef(ref z);
Assert(z == 5);

// No `ref` though on the `self` argument.
var c: C = {.x = 4};
c.NewMethod();
Assert(c.Get() == 7);
```

> **Note:** It is important that this restriction is syntactic, not just
> semantic, because it means that `ref` is never the first token of a full
> expression, and so we know without lookahead that a `ref` in a pattern context
> must be the start of a binding pattern, not the start of an expression
> pattern.

Normally an argument to a non-`ref` parameter should not be marked `ref`, but it
is allowed in a generic context where the parameter may sometimes be `ref`.

> Operators will mostly not take `ref` parameters, with these exceptions:

Reviewer 2: Would it be worth distinguishing *statement* operators from *expression* operators here?

I think most statement operators *do* take `ref` parameters, but for expression operators it is quite rare (just the first three in your list), and each of them is a deeply "builtin" thing that is for me fairly unsurprising to be "special", while assignment and increment seem much less "builtin".

Author: Split out the two cases.

retain a great deal of implementation flexibility around layout, and the
specifics of how they are lowered.

This specifically means we will need to incorporate `ref` bindings into the
`Call` interface and we will be adding complexity there that will need to be
handled by overloading. The changes to the `Call` interface is future work, and
overloading, once we add support, will need to carry additional complexity to
handle `ref`.

### Interaction with `returned var`

The rule is: `returned var` may only be used when there is a single component to
the return form, and it is default `->`.

```carbon
fn F(...) -> V {
  returned var v: V = ...;
  // ...
  return var;
}

fn F(...) -> {-> .a : T} {
  returned var ret: T = ...;
  // ...
  return var;
}
```

We can revisit and expand this later if this does not handle use cases we would
like to support.

### Use case: `Deref` interface

To support customization of the prefix-`*` dereferencing operator, we introduce
the `Deref` interface.

```carbon
interface Deref {
  let Result:! type;
  fn Op[bound ref self: Self]() -> ref Result;
}

final impl forall [T:! type] T* as Deref {
  where Result = T;

>   fn Op[bound ref self: Self]() -> ref T

Reviewer 2: Could the impl here drop the `ref`? (I understand why the interface needs it)

Author: Done

fn Check(y: Four) {
  Assert(y.Get(2) == 5);
}
Check(x);
```

**Future work**: this will in the future often be done with an overloaded
method, as in:

```carbon
class Four {
  overload Access {
    fn [bound ref self: Self](i: i32) -> ref i32 {
      Assert(i >= 0 and i < 4);
      return self.m[i];
    }
    fn [self: Self](i: i32) -> i32 {
      Assert(i >= 0 and i < 4);
      return self.m[i];
    }
  }
  private var m: array(i32, 4);
}

var x: HasMember = {.m = (0, 2, 4, 6)};
x.Access(2) += 1;
fn Check(y: Four) {
  Assert(y.Access(2) == 5);
}
Check(x);
```

This may be a common enough use case that we will want to introduce a dedicated
syntax:

```carbon
class HasMember {
  fn Access[bound ref? self: Self](i: i32) -> ref? i32 {
    Assert(i >= 0 and i < 4);
    return self.m[i];
  }
  private var m: array(i32, 4);
}
```

### Type completeness

Not a change by this proposal, but note that our existing rules will require the
type in a `ref` binding to be complete in situations where it would not need to
be if you were using a value binding with a pointer type instead. We may need to
change this in the future to match C++ which treats reference types like pointer

> types for completeness purposes.

Reviewer 2: My understanding is that while we'll need completeness before we would for a pointer, much like a value binding for a type, declarations of the binding still wouldn't require completeness? Worth mentioning that the result here is expected to have the same level of incompleteness support in ref bindings as value bindings?

Author: I think the text is technically correct as written. I can understand that it might not be as clear as it could be, but I'm having trouble figuring out how to address the points you make without introducing more confusion.

The situation is:

- The type in bindings is sometimes required to be complete, for example in definitions but some cases for declarations as well.
- `T*` is considered complete even when `T` is not.
- It is true that `T` in a value binding will generally be required to be complete in the same situations as a `ref` binding. However, a value binding for `T` is not a substitute for a reference binding for `T`, while a value binding for `T*` is much more so.

Author: I have added some text.

and on
[2025-05-12](https://docs.google.com/document/d/1Yt-i5AmF76LSvD4TrWRIAE_92kii6j5yFiW-S7ahzlg/edit?tab=t.0#heading=h.1mjh6unumnwu),
most notably:

-   **in place**: This convention was like `->`, but always using the "in place"
    convention where the caller allocates storage and provides the callee with
    its address, the callee initializes the storage at that address, and the
    caller is responsible for destroying after the return.
-   **var without storage**: The callee returns a pointer to the storage of a
    subobject of a `bound var` parameter, that caller is then responsible for
    destroying. A call to this function is reference expression, but with
    additional responsibility to destroy.
-   **hybrid**: If the type has a copy value representation or trivial
    destructive move then return the object representation directly; otherwise
    caller passes a pointer and callee initializes it.

There were also some variations on what the conditions for returning in
registers using the default return convention.

We seriously considered "var without storage", but the fact that it couldn't

> reliably used to initialize a variable, particularly in the middle of an object,

Reviewer 2: ```suggestion
reliably be used to initialize a variable, particularly in the middle of an object,
```

```carbon
fn F(...) -> (-> V1, -> V2, -> V3) {
  returned var (v1: V1, v2: V2, v3: V3) = ...;
  // ...
  return var;
}
```

### Other syntax for compound return forms

We considered other options for the syntax of compound return forms on
[2025-05-13](https://docs.google.com/document/d/1Yt-i5AmF76LSvD4TrWRIAE_92kii6j5yFiW-S7ahzlg/edit?tab=t.0#heading=h.bdznj2d0by2g),
[#syntax in Discord on 2025-05-14](https://discord.com/channels/655572317891461132/709488742942900284/1372285365162872943),
and
[2025-05-14](https://docs.google.com/document/d/1Yt-i5AmF76LSvD4TrWRIAE_92kii6j5yFiW-S7ahzlg/edit?tab=t.0#heading=h.52tb7l2he343).
The option of omitting the `->` in each component did not distinguish tuples
from paren return forms sufficiently:

```carbon
-> (ref i32, var i32)
-> (bool, ref i32)

// Is this a single return of a tuple, or a triple
// return using the default return convention?
-> (i32, i32, i32)
```

We also considered an approach where compound return forms would start with
`->?`, but this raised concerns about what the meaning of that syntax would be
and whether we want to expose users to that in cases we might be able to avoid
it.

### `ref` parameters allow aliasing

> If requiring `ref` parameters ends up being too restrictive, we could instead

Reviewer 2: ```suggestion
If requiring `ref` parameters to be `noalias` ends up being too restrictive, we could instead
```

    reference expression in the callee.
    -   Unlike pointers, a `ref` binding can't be rebound to a different object.
    -   This replaces `addr`, and is not restricted to the `self` parameter.
    -   A `ref` binding, like a `let` binding, can't be used in fields of
        classes or structs.
    -   When calling functions, arguments to non-`self` `ref` parameters are
        also marked with `ref`.
-   The return of a function can optionally be marked `ref`, `let`, or `var`.
    These control the category of the call expression invoking the function, and
    how the return expression is returned.
    -   These may be mixed for functions returning parens or brace forms.
-   Any parameters whose lifetime needs to contain the lifetime of the return
    must be marked `bound`.
-   The address of a `ref` binding is `nocapture` and `noalias`.
-   We mark parameters of a function that may be referenced by the return value
    with `bound`.

## Problem

Reference bindings have come up multiple times:

-   as a better alternative to `addr self: Self*`;
-   for use in [lambda captures](/docs/design/lambdas.md);
-   to support nested bindings within a destructured `var`, see

>     [issue #5250](https://github.com/carbon-language/carbon-lang/issues/5250)

Reviewer 3: Lines 82 and 88 are intended to start with /docs/... but when I click them they take me to /ref/docs/... which is a 404

Not sure how to fix this without making the full links with the domain name included. It looks like it affects other links too, these are just the first two I noticed.

Author: Those links work if I use view mode on this file:
https://github.com/carbon-language/carbon-lang/blob/79d6ddb0b0d7357d705e35e3f42b2ccc0e897090/proposals/p5434.md

Reviewer 3: Ah excellent! I didn't realize there was a discrepancy between the different methods of rendering a markdown document. Thanks!

    optimization. Also see
    [LLVM's pointer aliasing rules](https://llvm.org/docs/LangRef.html#pointer-aliasing-rules).
-   Marking a pointer as not captured, to allow optimizations, was originally
    done with
    [LLVM's `nocapture` attribute](https://releases.llvm.org/11.0.0/docs/LangRef.html#parameter-attributes),
    which has become
    [`captures(none)` and `captures(ret: address, provenance)`](https://llvm.org/docs/LangRef.html#function-attributes),
    which is governed by
    [pointer capture rules](https://llvm.org/docs/LangRef.html#pointer-capture).
-   Clang allows C++ code to use the
    [`clang::lifetimebound` attribute](https://clang.llvm.org/docs/AttributeReference.html#id8)
    to mark parameters that may be referenced by the return value, in order to
    detect some classes of use-after-free memory-safety bugs.
-   [C++ has reference types](https://en.cppreference.com/w/cpp/language/reference).

## Proposal

### `ref` bindings

We introduce a new keyword `ref`. This may be added to a `:` binding to mark it
as binding to a reference expression, as in:

```carbon
fn F(ptr: i32*) {
  // A reference binding `x`.

>   let ref x: i32 = *ptr;

Reviewer 4: At a high level, `let` can be thought of as a const thing, and `var` as a mutable thing, though there are of course storage implications. But this breaks that mental model for developers, so maybe we could preserve that by using `ref x:` instead?

Author: This is the "Top-level `ref` introducer" alternative. I have taken the conservative approach that we can always add this later, but I'd be interested in hearing if this is widely believed to be an issue.

Reviewer: That mental model breaks in a lot of places already. For example, `fn F(x: i32)` declares an immutable `x` despite the absence of `let`, and `let (var x: i32, y: i32) = ...` declares a mutable `x` despite the presence of `let`. We can't make that model coherent (without much deeper changes), so I don't think we should try to accommodate it.

Reviewer: I'm realizing that `-> let` really encourages the mental model where `let` means immutable (or at least by-value), but as noted elsewhere, I think that means we should choose a different syntax. If we don't, then I don't know, maybe allowing `ref` as a statement introducer would be desirable as well.

Reviewer 2: I think this has been resolved with the move to `-> val T`, and marking resolved. Re-open if I'm wrong though!

reference expression or an expression that can be converted to one. The bound
durable reference expression must outlive the `ref` binding.

The address of a `ref` bound name gives the address of the bound object, so
`&x == ptr` above. The reference itself does not have an address, and unlike a
pointer can't be rebound to reference a different object.

We remove `addr`, and use instead use `ref` for the `self` parameter when an
object is required. Note that the type will change from `Self*` to `Self` in
this case.

```carbon
class C {
  // ❌ No longer valid.
  fn OldMethod[addr self: Self*]() {
    // Previously would dereference `self` in
    // the body of the method.
    self->x += 3;
  }

  // ✅ Now valid.
  fn NewMethod[ref self: Self]() {
    // Now `self` is a reference expression,
    // and is not dereferenced.
    self.x += 3;
  }

  // ✅ Other uses are unchanged.
  fn Get[self: Self]() -> i32 {
    return self.x;
  }

  var x: i32;
}
```

Potentially abbreviating the syntax further (to allow `ref self` as a short form
of `ref self: Self`) is left as future work.

The `ref` modifier is allowed on any `:` parameter pattern (as an alternative to

> `var`) or any `let` binding in a function body.

Reviewer: Taken literally, this seems to be saying that the `ref` modifier is allowed on a `var` parameter, but not on a `var` binding in a function body. Is that the intent? If so, why the difference? If not, maybe this should be something like:

```suggestion
The `ref` modifier is allowed on any `:` binding that's not inside a `var` pattern.
```

(As a terminology nit, I think it's better to avoid the term "`let` binding" because it's ambiguous about whether it includes bindings like `let var (x: i32, y: i32) = ...`)

```carbon
fn AddTwoToRef(ref x: i32) {
  x += 1;
  let ref y: i32 = x;
  y += 1;
}

// Equivalent to:
fn AddTwoToRef(ref x: i32) {
  x += 1;
  let y_ptr: i32* = &x;
  *y_ptr += 1;
}
```

We add support for `ref` and `var` in a
[struct pattern](/docs/design/pattern_matching.md#struct-patterns) when using
the shorthand `a: T` syntax for `.a = a: T`:

```carbon
let {var a: i32, ref b: i32} = ...;

// Now equivalent to:
let {.a = var a: i32, .b = ref b: i32} = ...;
```

> The `ref` modifier is forbidden on the bindings in `class` or `struct` fields.

Reviewer: ```suggestion
The `ref` modifier is forbidden on the bindings in `class` fields.
```

Author: `ref` is allowed on brace forms, but not struct types. How should I write that?

Reviewer: Ah, sorry, I was mixing up struct types and struct patterns. Withdrawn.

```
class Invalid {
  // ❌ Invalid.
  ref var m: i32;
}
```

When calling functions, arguments to non-`self` `ref` parameters are also marked

> with `ref`. Continuing the example:

Reviewer: ```suggestion
In a function argument list, arguments to non-`self` `ref` parameters are also marked
with `ref`. Continuing the example:
```

It's important that this restriction is syntactic, not just semantic, because it means that `ref` is never the first token of a full expression, and so we know without lookahead that a `ref` in a pattern context must be the start of a binding pattern, not the start of an expression pattern.

```carbon
var z: i32 = 3;
AddTwoToRef(ref z);
Assert(z == 5);

// No `ref` though on the `self` argument.
var c: C = {.x = 4};
c.NewMethod();
Assert(c.Get() == 7);
```

Normally an argument to a non-`ref` parameter should not be marked `ref`, but it
is allowed in a generic context where the parameter may sometimes be `ref`.

Operators will mostly not take `ref` parameters, with these exceptions:

-   [the address-of operator](/docs/design/expressions/pointer_operators.md)
    `&`;

> -   [the indexing operator](/docs/design/expressions/indexing.md) `[`...`]`;

Reviewer: ```suggestion
-   the first operand of [the indexing operator](/docs/design/expressions/indexing.md) `[`...`]`;
```

> -   [the member access operator](/docs/design/expressions/member_access.md) `.`;

Reviewer: As far as I can tell, this isn't modeled as a function call, so it doesn't take parameters at all, `ref` or otherwise.

Author: See #3720 . Looking that over, looks like I should included an update to the `BindToRef` interface in this proposal.

Author: I've updated the text to make this more clear.

    and
-   [the assignment operators](/docs/design/assignment.md) such as `=`, `+=`,

>     and `++`.

Reviewer: ```suggestion
-   the left-hand operand of [assignment operators](/docs/design/assignment.md) such as `=` and `+=`.
-   [the `++` and `--` operators](/docs/design/assignment.md).
```

In the cases that they do, observe that the `ref` parameter will be the `self`
parameter, and so will not be marked with `ref` at the call site.

As an _experiment_, we are saying a pointer formed by taking the address of a

> `ref` bound name is LLVM-`captures(none)` and LLVM-`noalias`. This means that

Reviewer: Is there LLVM documentation we can link to for definitions of these terms?

Author: They are linked in the background section.

while a `ref` parameter could be passed into a function by address, the
restrictions also allow a "move-in-move-out" approach (once we define the move
operation), assuming it is not [marked `bound`](#bound-parameters). The intent
here is to leave the door open to a calling convention using registers and less
indirection for small-enough objects.

This means that the following code is invalid:

```carbon
fn F(ref a: i32, ref b: i32) -> bool;

fn G() -> bool {
  var v: i32 = 1;
  return F(ref v, ref v);
}
```

> ### `ref`, `let`, and `var` returns

Reviewer: It's not clear how `-> auto` and `=>` should work now: do they deduce only the type, or do they also deduce the form? Are there separate syntaxes for type-only deduction and form deduction?

Some if not all of that is probably out of scope for this proposal, but they should at least be flagged as open questions.

Author: I have added bullets to document how they are expected to behave.

The return of a function can optionally be marked `ref`, `let`, or `var`. These
control the category of the call expression invoking the function, and how the
return expression is returned.

```carbon
var global: i32 = 2;
fn ReturnRef() -> ref i32 {
  // ❌ Invalid: return 2;

  // ✅ Valid: return a reference expression with
  // sufficient lifetime.
  return global;
}
// Call `ReturnRef` and use the resulting reference.
ReturnRef() += 3;
Assert(global == 5);

// Result of `ReturnRef` can be bound using a `ref`
// binding.
fn AddFive() {
  let ref r: i32 = ReturnRef();
  r += 5;
}
AddFive();
Assert(global == 10);

fn ReturnLet() -> let i32 {
  return 2;
}
// ReturnLet() is a value expression.
let l: i32 = ReturnLet();

fn ReturnVar() -> var i32 {
  return 2;
}
// ReturnVar() is an initializing expression.
var j: i32 = ReturnVar();
```

> -   A call to a function declared `-> ref T` is a reference expression. The

Reviewer: ```suggestion
-   A call to a function declared `-> ref T` is a durable reference expression. The
```

>     function will return the address of a `T` object.

Reviewer: Could this say "The generated code for the function will..." or something like that, to be explicit that we're speaking about a different abstraction level than usual?

Author: Done.

-   A call to a function declared `-> let T` is a value expression. The function
    will return the value representation of `T`. Since values have no address,
    the value representation may be returned in registers.
-   A call to a function declared `-> var T` is an initializing expression. The
    caller provides the address of storage to initialize with an object of type
    `T`, that the caller owns upon return. The object will never be returned in

>     registers.

Reviewer: This seems confusingly understated: the object will not be returned _at all_. Maybe instead something like "The generated code for the function will not return a value"?

Author: Bullet deleted based on Chandler's suggestion.

>     registers.

Reviewer 2: After chatting with @zygoloid, I think I convinced him that we could fold together `-> T` and `-> var T`. I'm happy to write up more fully the idea, or discuss it / take notes on it, but leaving a quick comment here so I don't forget.

For my own memory, the most convincing thing was to start from thinking about `var` *parameters*, and how we would want those to work for an owning pointer like `std::unique_ptr`.

More details when I have a bit more time at a keyboard...

Author: I've updated the proposal to remove `-> var T`, and added it instead to the alternatives considered.

-   A call to a function declared `-> T` is unchanged. It is an initializing
    expression. It will either match `-> var T` or return by copy, depending on
    the initializing representation of `T`.

A function may have multiple returns, each with their own marker, by using a
paren or brace compound return form.

```carbon
fn ParenReturn()
    -> (-> bool, ->let f32, ->var C, ->ref i32) {
  return (true, 1.0, {.x = 3}, global);
}

fn BraceReturn()
    -> {-> .a: bool,
        ->let .b: f32,
        ->var .c: C,
        ->ref .d: i32} {
  return {.a = true, .b = 1.0,
          .c = {.x = 3}, .d = global};
}
```

If the return of a function may reference the storage of one or more parameters
to the function, those parameters must be marked `bound`. This allows the

> compiler to diagnose if the function's return is used after the lifetime of and

Reviewer: ```suggestion
compiler to diagnose if the function's return is used after the lifetime of the
```

[`clang::lifetimebound` attribute](https://clang.llvm.org/docs/AttributeReference.html#id8).

```carbon
fn Member(bound ref c: C) -> ref i32 {
  return c.x;
}

// Lifetime of a pointer includes the lifetime
// of what it points to.
fn Deref(bound p: i32*) -> ref i32 {
  return *p;
}

fn Both(bound pc: C*) -> ref i32 {
  return p->x;
}

fn Invalid1() -> ref i32 {
  var x: i32 = 4;
  // ❌ Error: returning reference to `x`
  // whose lifetime ends when this function
  // returns.
  return x;
}

fn Invalid2() -> ref i32 {
  var c: C = {.x = 1};
  // ❌ Error: returning reference bound to `c`
  // whose lifetime ends when this function
  // returns.
  return Member(c)
}
```

The address of a `bound ref` parameter is `captures(ret: address, provenance)`
instead of `captures(none)`.

## Details

The intent is that we would encourage using references instead of pointers when
possible. Their benefits are related to their limitations, so to get those
benefits we should use them when a use is restricted enough to be within those
limitations.

### Initializing returns

The `-> var T` return form guarantees that `returned var` is the same variable

> in the caller and callee. In contrast, the `-> T` return form that also produces

Reviewer: ```suggestion
in the caller and callee. In contrast, the `-> T` return form also produces
```

an initializing expression, but may introduce a copy depending on the
initializing representation. For example:

```carbon
fn F() -> var T {
  returned var v: T = {.p = &v};
  return var;
}

// Changing this to fn G() -> var T has a semantic effect
// if T has a copy initializing representation (and this
// passes safety checks).
fn G() -> T {
  return F();
}

fn H() -> var T {
  return G();
}

var x: T = H();

> // Is x.p == &x at this point?

Reviewer: "Has a semantic effect" is vague, and it seems unhelpful to ask but not answer the question at the end. Consider instead something like this:

```suggestion
fn G() -> var T {
  return F();
}

fn H() -> var T {
  return G();
}

var x: T = H();
// x.p == &x at this point, but might not be the case if
// the signature of `G` were `fn G() -> T` (depending
// on the initializing representation of T).
```

```

### Compound return forms and patterns

Mirroring the [paren](/docs/design/pattern_matching.md#tuple-patterns) and
[brace](/docs/design/pattern_matching.md#struct-patterns) pattern forms, we also

> support paren and brace return forms. Every element of these forms starts with

Reviewer: Can you remind me why you don't want to call these "tuple [return] forms" and "struct [return] forms"? Do you envision renaming any of the terms "tuple literal", "tuple pattern", or "tuple type" to say "paren" instead (or similar changes for "struct"->"brace")?

Author: I am trying to consistently use "paren" and "brace" to refer to forms and "tuple" and "struct" to refer to types. I am very concerned about keeping them separate since they are easy to confuse but have important differences.

Reviewer: But where does that leave "tuple literal" and "tuple pattern", which are neither types nor forms? Tuple literals in particular seem easy to confuse with tuple types, and possibly also with paren forms. Also, why would "tuple forms" and "tuple types" not be sufficiently distinguished by the fact that they have different head nouns?

Author: Since those operate in the domain of forms, I think it would make sense to rename those "paren literals" and "paren patterns." I can understand people disagreeing, though, and perhaps this distinction isn't one that will remain important to carry forward. In this proposal, though, it is very helpful to be able to distinguish the two different cases, particularly since you can return both tuple types and paren forms.

Reviewer 2: I think all of this uses the terminology "tuple forms" and "struct forms" now, so I think this thread is resolved and marking it as such. Do re-open if I've misunderstood. I also know that there is some friction with either naming pattern, but we seem to have settled on a consistent one and that seems enough for this iteration of proposals.

`->` and the kind of return (if not default).

```carbon
// Paren return form
fn ParenReturn(...)
    -> (-> bool, ->let f32, ->var C, ->ref i32);

let (var a: bool, b: f32, var c: C, ref d: i32)
    = ParenReturn(...);

fn BraceReturn(...)
    -> {-> .a: bool,
        ->let .b: f32,
        ->var .c: C,
        ->ref .d: i32};

// Binds to the names `x`, `y`, `z`, `w`:
let {.a = var x: bool,
     .b = y: f32,
     .c = var z: C,
     .d = ref w: i32} = BraceReturn(...);

// Binds to the names `a`, `b`, `c`, `d`:
let {var a: bool,
     b: f32,
     var c: C,
     ref d: i32} = BraceReturn(...);

// Above two can be mixed, binding to
// names `a`, `b`, `z`, `w`.
let {var a: bool,
     b: f32,
     .c = var z: C,
     .d = ref w: i32} = BraceReturn(...);
```

Note that we can distinguish between returning a compound form and a tuple or
struct with one token of lookahead. Only types are allowed after a `-> let`,
`-> var` or `-> ref`, not a compound return form. Examples:

```carbon
// Returns a tuple of type
// `(bool, f32, C, i32)`.
fn TupleReturn(...) -> (bool, f32, C, i32);

// ❌ Invalid, every element must start with

> // `->` to be a form.

Reviewer: ```suggestion
// ❌ Invalid, every element of a compound return form must
// start with `->`.
```

fn Invalid1(...) -> (bool, ->let f32);

// ❌ Invalid, can only specify return category
// on innermost `->`.
fn Invalid2(...) -> var (-> bool, -> f32);
```

To nest, use the default `->` in the outer form, as in:

```carbon
fn BracesInParens(...)

>     -> (->{-> .a: bool, ->let .b: f32}, ->var C, -ref i32);

Reviewer: Could we instead say that `->` always introduces an atomic form, so that this would be

```suggestion
fn BracesInParens(...)
    ({-> .a: bool, ->let .b: f32}, ->var C, -ref i32);
```

That way we don't need lookahead, but we avoid the noisy redundancy of the repeated arrows.

Author: So not starting with the first `->` introduces ambiguity without unbounded lookahead, since the explicit parameter list in parens is now optional. The later `->`s avoid ambiguity between types and return forms, without having to scan forward through an arbitrary number of opening brackets. It seems doable to have another parsing mode for return expressions, but my initial approach was to have something easy to parse since I did not expect this case to come up much in practice. I do agree that there are more arrows in this than is desirable. I'm just not sure if it is worth the parser complexity to change.

Reviewer 2: I think all of this is handled by the syntax update we landed on at the end here, so marking as resolved. Please re-open if I've missed something though.

// or without renaming:
let ({var a: bool, let b: f32},
     var c: C, ref d: i32) = BracesInParens();

fn ParensInBraces(...)
    -> {-> .a: bool,
        -> .b: (->let f32, ->var C),
        ->ref .c: i32};

let {a: bool,
     .b = (x: f32, var y: C),
     ref c: i32} = ParensInBraces(...);
```

This feature is intended to support cases like `enumerate` that will want to
return a value for the index but a reference to the element of the sequence
being enumerated.

### Nested binding patterns

Since a `ref` binding may only bind to a durable reference expression, it can't
be used to bind the result of a function returning an initializing expression.
However, if the initializing expression is bound to a `var`, any nested patterns
are reference binding patterns bound to the subobject, following
[proposal #5164: "Updates to pattern matching for objects"](https://github.com/carbon-language/carbon-lang/pull/5164).

For example:

```carbon
fn F() -> (bool, (C, i32));
let (b: bool, var (c: C, i: i32)) = F();
```

is equivalent to:

```carbon
fn F() -> (bool, (C, i32));
let (b: bool, var v: (C, i32)) = F();
let ref c: C = v.0;
let ref i: i32 = v.1;
```

> 

Reviewer: Maybe explicitly spell out here that this is why `ref` is disallowed inside `var`: it would be redundant.

Author: Added a note.

It is an error to return a reference to a temporary object that won't live once
the function returns, even if it is a parameter marked `bound`.

```carbon
fn Invalid1() -> i32* {
  var x: i32 = 4;
  // ❌ Invalid.
  return &x;
}

fn Invalid2(bound x: i32) -> ref i32 {
  var y: i32 = x;
  // ❌ Invalid.
  return y;
}

// ✅ Valid
fn Valid1(bound p: i32*) -> ref i32 {
  return *p;
}

fn Invalid3(bound var x: i32) -> ref i32 {
  // ❌ Invalid: lifetime of `var` parameter
  // ends when function returns.
  return x;
}

class ReturnMember {
  // ✅ Valid
  fn ValidRef[bound ref self: Self]() -> ref i32 {
    return self.m;
  }

  // ❌ Invalid: can't return reference to value.
  fn InvalidLet[bound self: Self]() -> ref i32 {
    return self.m;
  }

  // ❌ Invalid: `var self` lifetime ends.
  fn InvalidVar[bound var self: Self]()
      -> ref i32 { return self.m; }

  var m: i32;
}

class DerefPointerMember {
  // ✅ Valid
  fn ValidRef[bound ref self: Self]() -> ref i32 {
    return *self.pm;
  }

  // ✅ Valid
  fn ValidLet[bound self: Self]() -> ref i32 {
    return *self.pm;
  }

  // ✅ Valid
  fn ValidVar[bound var self: Self]()
      -> ref i32 { return *self.pm; }

  var pm: i32*;
}
```

> The `bound` parameters must include everything referenced by the return, but

Reviewer: That doesn't seem quite right, since the return can reference global variables. The phrasing is also a little awkward. How about something like "The returned expression for a `-> ref` cannot reference any local names not marked with `bound`"?

Author: That isn't a strictly accurate formulation, since you can reference a local name that shares storage with something bound. I've changed the text to add globals.

need not be referenced, particularly not on every code path.

```carbon
// Result references `r` if `b` is true, and `p`
// otherwise. Valid as long as both are marked `bound`.
fn Conditional(b: bool, bound ref r: C, bound p: C*)
    -> ref C {
  if (b) {
    return r;
  } else {
    return *p;
  }
}
```

The parameters of functions defined in an interface may also be marked as

> `bound`. The `impl` of that interface for a type can only have `bound` in a
> subset of the positions present in the interface.

Reviewer: The "can only have `bound` in a subset" phrasing is a little hard to follow. How about this?

```suggestion
`bound`. The `impl` of that interface for a type can omit occurrences of
`bound` from the interface, but cannot add new ones.
```

lifetime of temporaries is [future work](#temporary-lifetimes).

Note that all combinations of a `var`/`ref`/`let` return can be bound to a
`var`/`ref`/`let` parameter. Examples:

```carbon
fn RefToLet(bound ref x: C) -> let D { return x.d; }
fn LetToRef(bound y: C) -> ref D { return *y.ptr; }
fn VarToRef(bound var p: i32*) -> ref i32 { return *p; }
fn VarToVar(bound var p: i32*) -> var i32* { return p; }
```

For full safety, we need each bound parameter to be immutable for the duration
of the lifetime of the returned result. However, the objective for now is only
matching `[[clang::lifetimebound]]`, which has the goal of preventing some
classes of bugs, not full memory safety. We will reconsider this with the memory
safety design.

### How addresses interact with `ref`

The address of a `ref` binding is `noalias` and either `captures(none)` or
`captures(ret: address, provenance)`, depending on whether the binding is marked
`bound`.

-   `noalias` means like C `restrict`; you can't observe mutations through
    aliases; mutation through a restricted pointer is not observable through
    another pointer

> -   `captures(none)` means no transitive escape, can pass a nocapture pointer to
>     another nocapture function, can't store to memory or return

Reviewer: The omitted words in this sentence are hard to reconstruct. Is this right? If so I suggest making that change or something like it.

```suggestion
-   `captures(none)` means there is no transitive escape: you can pass a nocapture pointer to
    another nocapture function, but you can't store to memory or return
```

-   `captures(ret: address, provenance)`: is like `captures(none)` but may be
    referenced by a return.

> The `noalias` semantics are the minimum for the optimization. But this condition

Reviewer: Which optimization?

Author: Added "move-in-move-out"

that the `nocapture` and `noalias` modifiers don't match C++ restrictions,
particularly on the `this` parameter that we are going to require migrate to
`ref self`. We may have to add back in `addr` to allow a different pointer type
for those cases.

### Part of the expression type system, not object types

Much like value/`let` and `var` bindings, `ref` binding and the new return forms
are are part of the type system, but only through expression categories,
patterns (function parameters and so on), and returns. Specifically, we don't
expect them to be part of the _object types_ in Carbon. Like value bindings, we
retain a great deal of implementation flexibility around layout, and the
specifics of how they are lowered.

This specifically means we will need to incorporate `ref` bindings into the
`Call` interface and we will be adding complexity there that will need to be
handled by overloading. The changes to the `Call` interface is future work, and
overloading, once we add support, will need to carry additional complexity to
handle `ref`.

### Interaction with `returned var`

The rule is: `returned var` may only be used when there is a single component to

> the return form, and it is either `->var` or default `->`.

Reviewer 2: I wonder -- could we instead simply say that `returned ...` must match the structure of the return pattern, and any non-`->var` or non-`->` components must be initialized in the declaration?

(also OK deferring this if folks prefer, it just seems a clean generalization)

Author: This was the first option in the "`return var` with compound return forms" alternative considered. How would you expect that example to be written?

```
fn F(...) -> (->ref R, ->let L, -> V) {
  returned (ref r: R, let l: L, var v: V) = (...);
  // or?
  returned let (ref r: R, let l: L, var v: V) = (...);

  // What goes here?
  return ???;
}
```

I also don't see a way to avoid initializing all of them together using this approach.

Reviewer 5: We could shorten `return var;` to simply `return;` when a `returned ...` declaration is in scope. I'm not sure how much clarity we're getting from including the `var` keyword there. If we do that, we might also want to consider allowing flowing off the end of a function with a `returned` declaration in its outermost block scope.

reference captures.

### No call-site annotation

This question was discussed on
[2025-05-07](https://docs.google.com/document/d/1Yt-i5AmF76LSvD4TrWRIAE_92kii6j5yFiW-S7ahzlg/edit?tab=t.0#heading=h.sfx9d7ltud5).

We decided that the marking is not about lifetime, but ability to mutate. A
`let` may reference an object in a similar way to a `ref`, restricting
operations on the original object, but we are not going to mark `let`s since
those restrictions are enforced by the compiler. We thought the ability to
mutate, though, was something important enough to highlight to readers of the
code, even at the expense of extra work for the writer.

Swift `inout` parameters are
[marked at caller with an `&` before the argument](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/functions/#In-Out-Parameters).
Jordon Rose has published
[a regret](https://belkadan.com/blog/2021/12/Swift-Regret-inout-Syntax/) that
they didn't use `inout` to mark the argument instead of `&`.

On the other hand, not marking is not known to be a source of bugs.

This is a "try it and see how well it works" sort of decision.

> ### Top-level `ref` introducer

Reviewer 3: Just got finished reading! Really excited about this direction.

One question comes to mind:

> Top-level ref introducer
> 
> For now, we don't believe let ref to be so common as to need a shorter way to write, unlike what we do for var.

There doesn't appear to be much written on this. Why was this decision made? I personally find it surprising that the following code is invalid:

```
fn F(ptr: T*) {
  ref t: T = *ptr;
}
```

Is this something that we have strong consensus on? Or is it possible to reconsider before landing the proposal?

Thanks!

Author: I don't know how strong the consensus, but I think the reason here is enough for my personal opinion to be to go forward with the approach in the proposal.

Author: Leads issue #5523 covers this question.

Reviewer 2: (Marking as resolved, this seems to have a good body of rationale now.)

    object.
-   A call to a function declared `-> let T` is a value expression. The function
    will return the value representation of `T`. Since values have no address,
    the value representation may be returned in registers.
-   The behavior of a call to a function declared `-> T` is unchanged. It is an
    initializing expression, returning in place or by copy depending on the
    initializing representation of `T`.
-   The behavior of `auto` as the return type is unchanged, but now supports an
    optional `ref` or `let` between the `->` and `auto`. `-> auto` continues to
    return an initializing expression, `-> let auto` returns a value expression,
    and `-> ref auto` returns a durable reference expression.
-   Using `=>` to specify a return continues to return an initializing
    expression, as before. Trying to infer the expression category from the
    category of the expression after the `=>` runs into the problem of this
    often requiring the parameters to be marked `bound`. A more complicated
    rule, for example using whether any parameter is marked `bound`, could be
    adopted in the future, if the simple rule proves inadequate.

> A function may have multiple returns, each with their own marker, by using a
> paren or brace compound return form.
> 
> ```carbon
> fn ParenReturn()
>     -> (->let bool, ->ref i32, -> C) {
>   return (true, global, {.x = 3});
> }
> 
> fn BraceReturn()
>     -> {->let .a: bool,
>         ->ref .b: i32,
>         -> .c: C} {
>   return {.a = true,
>           .b = global,
>           .c = {.x = 3}};
> }
> ```

Reviewer 2: I was struggling a bit with the heavy usage of the `->` symbol in these, and I discussed an alternative syntax with @josh11b in the open discussion that seemed promising.

Discussion: https://docs.google.com/document/d/1Yt-i5AmF76LSvD4TrWRIAE_92kii6j5yFiW-S7ahzlg/edit?tab=t.0#heading=h.52ru7ner80b4

Rough sketch of the syntax adjustment from that discussion:

```
// `->` begins a "return form", whatever the right word
fn F()     -> <return-form>;

// Within any return form, there are N syntactic overrides where the thing following
// the arrow *does not* form a type expression
fn Let()   -> let <type-expr>
fn Ref()   -> ref <type-expr>
fn Var()   -> var <type-expr>
fn Paren() -> ( <return-form>, ... )
fn Brace() -> { .<id>: <return-form>, ... }

// Otherwise, implicit `var`:
fn Other() -> <type-expr>
```

Basically, `(` and `{` are always compound return forms unless they are *explicitly* placed in a type expression using `let`, `ref`, or `var`. This would make compound returns the *default*, which is a not-small change. But as I have thought about this proposal and the consequences, I think it is a reasonable if not preferable default, and it provides a nice way to organize the syntax.

There is also a minor change to put the modifier `let` or `ref` after the `:` in the brace form. I don't feel strongly about that either way, but seemed slightly more consistent this way as this `:` is a very different kind of `:`.

With this, I think these examples would change as follows:
````suggestion
A function may have multiple returns, each with their own marker, by using a
paren or brace compound return form.

```carbon
fn ParenReturn() -> (let bool, ref i32, C) {
  return (true, global, {.x = 3});
}

fn BraceReturn()
    -> {.a: let bool,
        .b: ref i32,
        .c: C} {
  return {.a = true,
          .b = global,
          .c = {.x = 3}};
}
```
````

Reviewer 2: A slight amendment to this from open discussion with @zygoloid and @josh11b --

Whenever we have a paren form or brace form with all *implicit* <type-expr> (IE, no `let`, `ref`, or `var` keywords), we don't form a compound return, the paren or brace form itself is just a `<type-expr>` and the var goes outside of it. This basically pushes the implicit `var` to the outer most layer it can be.

This gives it the important property that a type `T` that is maybe an alias for a tuple type and an explicitly written tuple type as `(i32, i32)` have the same behavior -- they both are var, neither is `(var i32, var i32)`. If the user wants explicitly separate return forms despite all the forms being `var`, they can use the explicit `var` keyword to achieve this.

Author: Updated.

are are part of the type system, but only through expression categories,
patterns (function parameters and so on), and returns. Specifically, we don't
expect them to be part of the _object types_ in Carbon. Like value bindings, we
retain a great deal of implementation flexibility around layout, and the
specifics of how they are lowered.

This specifically means we will need to incorporate `ref` bindings into the
`Call` interface and we will be adding complexity there that will need to be
handled by overloading. The changes to the `Call` interface is future work, and
overloading, once we add support, will need to carry additional complexity to
handle `ref`.

### Interaction with `returned var`

The rule is: `returned var` may only be used when there is a single component to
the return form, and it is either `->var` or default `->`.

```carbon
fn F(...) -> var V {
  returned var v: V = ...;
  // ...
  return var;
}

fn F(...) -> T {
  returned var ret: T = ...;
  // ...
  return var;
}

fn F(...) -> {->var .a : T} {
  returned var ret: T = ...;
  // ...
  return var;
}
```

We can revisit and expand this later if this does not handle use cases we would
like to support.

> ### Use case: `Deref` interface

Reviewer: (Sorry, I forgot to include this comment in my first batch.)

When a proposal is making changes to an already-documented design, I think we should err pretty strongly on the side of making those documentation changes as part of the proposal. Seeing the changes in context should make it substantially easier to review, and may surface important issues that should be considered as part of the proposal (plus, it eliminates the time window where the design docs no longer reflect the adopted design).

This section and the next one seem like they would particularly benefit from this, but I think it applies to this proposal as a whole.

Author: I have updated `expressions/indexing.md` with the changes from the next section, but `Deref` is not currently in the design.

As for updating the rest of the design, I think it is important to get consensus on the proposal first, otherwise iteration will be significantly more expensive. Updating all of the design docs to reflect this change is going to be a big effort which we will need to prioritize against other work, and may need to be done only partially at first.

Reviewer 2: FWIW, I agree that it doesn't make sense to try to do a *complete* update in the proposal. I appreciate the binding and indexing updates, and think that's roughly the right amount of design doc updating for proposals. And I think we have a new PR to discuss articulating this expectation a bit better, but I don't think it needs to block this proposal.

The rationale to add `ref` instead of staying with pointers was discussed on
[2025-05-01](https://docs.google.com/document/d/1Yt-i5AmF76LSvD4TrWRIAE_92kii6j5yFiW-S7ahzlg/edit?tab=t.0#heading=h.dffumsu6wzlc).
In addition to the motivating problems given in
[the "Problem" section](#problem), that discussion included some additional
depth to the reasons to add reference bindings.

There is a tension between wanting to have mutating expressions and only having
pointers. You need some concept like a reference in order to mutate an object
with an expression. The question is how small a box the references are
restricted to, and where the line is drawn. C has lvalues, which contain
references but are restricted to a quite small box. Reference bindings
specifically are about keeping a small box around references while still adding
enough expressivity to support our use cases. We have started with a model
similar to C, but it fell down when it comes down to composition. Decomposing an
expression into pieces loses the tools the expression provided to you. The
missing tool for that was reference bindings.

We saw how much we were leaning on value bindings. The asymmetry between having

> value binding but not referenced bindings when have value expressions and

Reviewer 2: ```suggestion
value binding but not reference bindings when have value expressions and
```

registers using the default return convention.

We seriously considered "var without storage", but the fact that it couldn't
reliably be used to initialize a variable, particularly in the middle of an
object, meant it did not seem valuable enough to include.

It seemed more valuable to support the "in place" return convention. That return
form allows you to guarantee knowing the address of the object being
constructed, and was a good match for `returned var`. However, we realized that
`var` declarations shouldn't always be associated with in-memory storage, in
particular for types that may be trivially moved. For example, a `var` parameter
with the C++ type `std::unique_ptr<T>` should be passed in registers. A function
returning a `std::unique_ptr<T>` in place would not be as efficient as returning
it by moving it into registers.

### `return var` with compound return forms

We considered various syntax options on
[2025-05-12](https://docs.google.com/document/d/1Yt-i5AmF76LSvD4TrWRIAE_92kii6j5yFiW-S7ahzlg/edit?tab=t.0#heading=h.1mjh6unumnwu),
but none of them seemed good enough to justify inclusion at this time:

```carbon

> fn F(...) -> (->ref R, ->val L, -> V) {
>   // No longer a `var` being returned. Ideally these
>   // shouldn't have to be initialized together.
>   returned ??? (ref r: R, val l: L, var v: V) = ...?
>   return var;
> }
> 
> // We could restrict to one `var` return component,
> // but this is a lot of machinery for a small increase
> // in expressiveness and applicability.
> fn F(...) -> (->ref R, ->val L, -> V) {
>   returned var v: V = ...;
>   let l: L = ...;
>   return (*r, l, var);
> }
> 
> fn F(...) -> (->ref R, ->val L, -> V) {
>   // These don't have the right category, and ideally
>   // shouldn't have to be initialized together.
>   returned var ret: (R, L, V) = ...?
>   return var;
> }
> 
> fn F(...) -> (->ref R, ->val L, -> V) {
>   returned var (_, _, var v: V) = <what goes here?>;
> }
> ```
> 
> There was another approach we considered for `returned var` originally:
> 
> ```carbon
> fn F(...) -> (->ref R, ->val L, -> v1: V1, -> v2: V2) {
>   // ...
>   // Must use the same names for `->` returns with bound names.
>   return (r, l, v1, v2);
> }
> ```
> 
> But this had downsides that still apply:
> 
> -   Requires `V1` and `V2` to have unformed states. Otherwise, `v1` and `v2`
>     would need be initialized when they are declared.
> -   This does not support only having some branches use `return var`.
> 
> Our current approach handles our main use case for `returned var`: factory
> functions.
> 
> We could support an "only `->`s" approach in the future if we want:
> 
> ```carbon
> fn F(...) -> (-> V1, -> V2, -> V3) {
>   returned var (v1: V1, v2: V2, v3: V3) = ...;
>   // ...
>   return var;
> }
> ```

Reviewer 2: Updating the syntax here and tweaking the discussion to ready a bit more clearly in the face of the syntax we landed on...

````suggestion
fn F(...) -> (ref R, val L, V) {
  // No longer a `var` being returned. Ideally these
  // shouldn't have to be initialized together.
  returned ??? (ref r: R, val l: L, var v: V) = ...?
  return var;
}

// We could restrict to one `var` return component,
// but this is a lot of machinery for a small increase
// in expressiveness and applicability.
fn F(...) -> (ref R, val L, V) {
  returned var v: V = ...;
  let l: L = ...;
  return (*r, l, var);
}

fn F(...) -> (ref R, val L, V) {
  // These don't have the right category, and ideally
  // shouldn't have to be initialized together.
  returned var ret: (R, L, V) = ...?
  return var;
}

fn F(...) -> (ref R, val L, V) {
  returned var (_, _, var v: V) = <what goes here?>;
}
```

There was another approach we considered for `returned var` originally:

```carbon
fn F(...) -> (ref R, val L, v1: V1, var v2: V2) {
  // ...
  // Must use the same names for the `var` (implicit or explicit) returns
  // with bound names.
  return (r, l, v1, v2);
}
```

But this had downsides that still apply:

-   Requires `V1` and `V2` to have unformed states. Otherwise, `v1` and `v2`
    would need be initialized when they are declared.
-   This does not support only having some branches use `return var`.

Our current approach handles our main use case for `returned var`: factory
functions.

We could support an "only `var`s" approach in the future if we want:

```carbon
fn F(...) -> (var V1, var V2, var V3) {
  returned (var v1: V1, var v2: V2, var v3: V3) = ...;
  // ...
  return var;
}
```
````

    -   [`=>` infers form, not just type](#-infers-form-not-just-type)

<!-- tocstop -->

## Abstract

-   A parameter binding can be marked `ref` instead of `var` or the default. It
    will bind to reference argument expressions in the caller and produces a
    reference expression in the callee.
    -   Unlike pointers, a `ref` binding can't be rebound to a different object.
    -   This replaces `addr`, and is not restricted to the `self` parameter.
    -   A `ref` binding, like a value binding, can't be used in fields of
        classes or structs.
    -   When calling functions, arguments to non-`self` `ref` parameters are
        also marked with `ref`.
-   The return of a function can optionally be marked `ref`, `val`, or `var`.
    These control the category of the call expression invoking the function, and
    how the return expression is returned.
    -   These may be mixed for functions returning tuple or struct forms.
-   The address of a `ref` binding is `nocapture` and `noalias`.
-   We mark parameters of a function that may be referenced by the return value
    with `bound`.

## Problem

> Reference bindings have come up multiple times:

Reviewer 5: Is there anything we should be saying about interop/migration here?

Author: I think I was a bit hesitant since with the restrictions on `ref` in this proposal it wasn't super clear whether this feature would be usable for interop&migration. However, it definitely is part of the motivation for introducing them, and having worked more on the safety story it looks like `ref` bindings will gain the flexibility to perform this role. I've added it here and as future work in the "Improved interop and migration with C++ references" details section.

[re-add `addr` back if needed](#restore-addr).

```carbon
class C {
  // ❌ No longer valid.
  fn OldMethod[addr self: Self*]() {
    // Previously would dereference `self` in
    // the body of the method.
    self->x += 3;
  }

  // ✅ Now valid.
  fn NewMethod[ref self: Self]() {
    // Now `self` is a reference expression,
    // and is not dereferenced.
    self.x += 3;
  }

  // ✅ Other uses are unchanged.
  fn Get[self: Self]() -> i32 {
    return self.x;
  }

  var x: i32;
}
```

Potentially abbreviating the syntax further (to allow `ref self` as a short form
of `ref self: Self`) is left as future work.

The `ref` modifier is allowed on any `:` binding that's not inside a `var`
pattern.

```carbon
fn AddTwoToRef(ref x: i32) {
  x += 1;
  let ref y: i32 = x;
  y += 1;
}

// Equivalent to:
fn AddTwoToRef(ref x: i32) {
  x += 1;
  let y_ptr: i32* = &x;
  *y_ptr += 1;
}
```

We add support for `ref` and `var` in a
[struct pattern](/docs/design/pattern_matching.md#struct-patterns) when using
the shorthand `a: T` syntax for `.a = a: T`:

```carbon
let {var a: i32, ref b: i32} = ...;

// Now equivalent to:
let {.a = var a: i32, .b = ref b: i32} = ...;

> ```

Reviewer 5: This takes us one step closer to `{` ambiguity. Previously we could distinguish between a struct literal/pattern and a non-empty block with only up to two tokens of lookahead (the struct cases start with `.` or `_` or identifier followed by `:`, and the block cases don't). Now we have things like:
```carbon
fn F() -> X { var a: i32 = 0; }
```
... where we're getting incrementally closer to ambiguity. We've got a few more steps before we get there (we don't have an `X{...}` expression yet, and `var ...` is only allowed in struct patterns rather than struct expressions), and I think we're still fine, but this is cutting down our options for future syntactic expansion a little.

Nonetheless this does seem like a sufficiently valuable feature that I think we should add it.

Author: Added this text to a note.

> The `ref` modifier is forbidden on the bindings in `class` or `struct` fields.

Reviewer 5: At least the struct part here seems to contradict the earlier

> The `ref` modifier is allowed on any `:` binding that's not inside a `var` pattern.

It might be clearer to integrate this into that earlier statement?

Author: Modified in both places.

```
var OuterSize:! i32 = 123;

class Invalid {
  // ❌ Invalid.

>   let ref Size:! i32 = OuterSize;

Reviewer 5: I'm not sure this is a good example of the previous rule -- if this were both valid without `ref` and valid outside a class, I think it should be valid with `ref` inside a class too. (I think it's invalid outside a class because we only said that `ref` can be used with `:`, not `:!`, and I think it's invalid without `ref` in a class because we don't say that we allow `let` declarations in classes at all.)

Author: Changed `:!` to `:` in both places. Looks like my fingers have gotten to used to typing `:!`.

Reviewer 5: Hm, isn't this still invalid because you can't have a `let` binding in a class? (If we allow a `let` in a class to declare a static member, I'd expect `let ref` to also be valid.) I *think* the mention of classes here is redundant, except if we want to point out that -- unlike C++ -- we don't have any way to introduce references as class members.

Reviewer 2: I've suggested an edit to hopefully clarify that this is more trying to document the intent of not (yet) having ref bindings as fields, despite there not being a syntactic way to really get there at this point.

Reviewer 5: Suggested edit works for me.

Author: Edit has been applied

-   [the member access operator](#use-case-member-binding-interfaces) introduced
    in
    [proposal #3720: "Member binding operators"](https://github.com/carbon-language/carbon-lang/pull/3720)
    `.`; and
-   the left-hand operand of [assignment operators](/docs/design/assignment.md)
    such as `=` and `+=`.
-   [the `++` and `--` operators](/docs/design/assignment.md).

Even in these cases, the arguments will not be marked with `ref` at the call
site. (Generally the `ref` parameter is the `self` parameter, and so wouldn't be
marked. The exception is `BindToRef`, but we don't want to mark its argument
with `ref`.)

As an _experiment_, we are saying a pointer formed by taking the address of a
`ref` bound name is LLVM-`captures(none)` and LLVM-`noalias`. This means that
while a `ref` parameter could be passed into a function by address, the
restrictions also allow a "move-in-move-out" approach (once we define the move
operation), assuming it is not [marked `bound`](#bound-parameters). The intent
here is to leave the door open to a calling convention using registers and less
indirection for small-enough objects.

> This means that the following code is invalid:

Reviewer 5: What do you mean by invalid here -- that it's undefined (which would be the consequence of the LLVM attributes you mentioned if `F` stores through one of the references and reads through the other) or that we will reject the program? The phrasing "This means..." suggests the former to me but I suspect your intent is the latter.

(I've not read the rest of the proposal yet, so you might cover this later, but in the latter case I'd want to see a description of the rule we're enforcing -- or future work to come up with one.)

Author: Enforcement is desired, but really part of the memory safety story so I'm going to make this erroneous behavior for now.

Added some text after the example.

  return (true, global, {.x = 3});
}

fn StructReturn()
    -> {.a: val bool,
        .b: ref i32,
        .c: C} {
  return {.a = true,
          .b = global,
          .c = {.x = 3}};
}
```

If the return of a function may reference the storage of one or more parameters
to the function, those parameters must be marked `bound`. This allows the
compiler to diagnose if the function's return is used after the lifetime of the
`bound` parameter ends. The semantics of `bound` are intended to match the
[`clang::lifetimebound` attribute](https://clang.llvm.org/docs/AttributeReference.html#id8).

```carbon
fn Member(bound ref c: C) -> ref i32 {
  return c.x;
}

// Lifetime of a pointer includes the lifetime
// of what it points to.
fn Deref(bound p: i32*) -> ref i32 {
  return *p;
}

fn Both(bound pc: C*) -> ref i32 {
  return p->x;
}

fn Invalid1() -> ref i32 {
  var x: i32 = 4;
  // ❌ Error: returning reference to `x`
  // whose lifetime ends when this function
  // returns.
  return x;
}

fn Invalid2() -> ref i32 {
  var c: C = {.x = 1};
  // ❌ Error: returning reference bound to `c`
  // whose lifetime ends when this function
  // returns.
  return Member(c)
}
```

> The address of a `bound ref` parameter is `captures(ret: address, provenance)`

Reviewer 5: It would have helped me to mention that `captures(...)` here is the LLVM attribute, not some new Carbon syntax that will be introduced later. (It's in the background but I'd forgotten already!)

Author: Added clarification and links.

also support tuple and struct return forms.

```carbon
// `->` begins a "return form"
fn F()     -> <return-form>;

// Within any return form, if the first token is
// `val`, `ref`, `var`, `(`, or `{`, it is not
// treated as type expression:

// Value return, with type as specified
fn Val()   -> val <type-expr>

// Reference return, with type as specified
fn Ref()   -> ref <type-expr>

// Initializing return, with type as specified
fn Var()   -> var <type-expr>

// Tuple compound return, with a list of
// return forms.
fn TupleCompound() -> ( <return-form>, ... )
// Tuple return, with a list of type
// expressions. Used if all members of the
// list are type expressions.
fn Tuple() -> ( <type-expr>, ... )

// Struct compound return, with a mapping from
// designators to return forms.
fn StructCompound()
    -> { .<id>: <return-form>, ... }
// Struct return, used if all of the members
// are type expressions.
fn Struct() -> { .<id>: <type-expr>, ... }

// Otherwise, implicit `var` means returns
// an initializing expression.
fn Other() -> <type-expr>
```

Note that in the absence of `val`, `ref`, and `var` keywords, the implicit `var`

> is placed in the outermost position, minimizing the number of returns. So

Reviewer 5: What does "minimizing the number of returns" here mean? Do you mean something like, minimizing the number of primitive forms returned?

Author: Yes.

// Contrast with a compound struct form containing
// a tuple type (not compound):
fn TupleInBraces(...)
    -> {.a: bool, .b: (f32, C), .c: ref i32};
// Equivalent to:
    -> {.a: var bool, .b: var (f32, C), .c: ref i32};
```

This feature is intended to support cases like `enumerate` that will want to
return a value for the index but a reference to the element of the sequence
being enumerated.

### Nested binding patterns

Since a `ref` binding may only bind to a durable reference expression, it can't
be used to bind the result of a function returning an initializing expression.
However, if the initializing expression is bound to a `var`, any nested patterns
are reference binding patterns bound to the subobject, following
[proposal #5164: "Updates to pattern matching for objects"](https://github.com/carbon-language/carbon-lang/pull/5164).

For example:

```carbon
fn F() -> (bool, (C, i32));
let (b: bool, var (c: C, i: i32)) = F();
```

is equivalent to:

```carbon
fn F() -> (bool, (C, i32));
let (b: bool, var v: (C, i32)) = F();
let ref c: C = v.0;
let ref i: i32 = v.1;
```

Note that `ref` is disallowed inside `var` since that would be redundant.

> ### Mututation restriction on objects bound to a value

Reviewer 5: It would have helped me to introduce `bound` before this section, given that presumably that's what makes it relevant for this proposal.

Also, is this proposal the one introducing the "erroneous behavior" rule you mention below, or are you just reminding the reader of it? I thought we already had that from a previous proposal (maybe #2006), and we [describe it in the design already](https://docs.carbon-lang.dev/docs/design/values.html#value-expressions:~:text=Once%20an%20object,instead%20just%20%E2%80%9Cerroneous%E2%80%9D.).

Author: The intention of this section is to be about value bindings, not `bound` (with the distinction in the section title only being conveyed by not using code font, which is indeed pretty subtle), but I guess this may end up being a restriction on `bound` as well?

Sorry, I thought we had already defined "erroneous behavior," but it appears we haven't. The "No optimization on erroneous behavior" section was intended to be a clarification, where I copied down what Chandler said recently on the topic. It looks like we have a need for a definition to be included in the design. Is there a good definition @chandlerc we can use? @chandlerc Should it be added by this proposal or a separate proposal?

Reviewer 5: I think [this section](https://docs.carbon-lang.dev/docs/design/values.html#value-expressions) already does define "erroneous behavior", albeit only implicitly.

Reviewer 2: Happy to look for something we can use, but I'm comfortable with that being a follow-up proposal to add a good definition. Also don't think its blocking given that we have an implicit one and the clarification needed in this proposal.

    under "as-if", and we don't need anything else.
-   We are deferring the decision of whether strict enforcement is enabled in
    Carbon's current C++-friendly mode, when not explicitly marking the code as
    "unsafe."

This was
[discussed 2025-05-22](https://docs.google.com/document/d/1Yt-i5AmF76LSvD4TrWRIAE_92kii6j5yFiW-S7ahzlg/edit?tab=t.0#heading=h.uot97ukynlsi)
and then made the subject of
[leads issue #5524](https://github.com/carbon-language/carbon-lang/issues/5524).

### No optimization on erroneous behavior

The Carbon compiler should not optimize on erroneous behavior, ever, unless the
compiler literally proves that it does not occur, ever. In which case, we don't
need any license to start optimizing on this, as it falls under "as-if".

The fact that undefined behavior ("UB",
[cppreference](https://en.cppreference.com/w/cpp/language/ub.html),
[wikipedia](https://en.wikipedia.org/wiki/Undefined_behavior)) provides "as-if"
_without_ a proof is precisely the risk of using UB for any semantics, and why
we don't use it here and elsewhere we use erroneous behavior.

### `bound` parameters

It is an error to return a reference to a temporary object that won't live once

> the function returns, even if it is a parameter marked `bound`.

Reviewer 5: The examples below show references to local variables, which I'd been thinking of as distinct from temporary objects. Could we use a term such as "local object" for temporary objects + local variables + `var` parameters?

Author: Done.

```carbon
fn Invalid1() -> i32* {
  var x: i32 = 4;
  // ❌ Invalid.

>   return &x;

Reviewer 5: This doesn't look like an example of the rule above. Should it say "reference or pointer" instead of just "reference"?

Reviewer 5: Also if we're extending this to pointers as well as references, rebindability becomes a problem. What rule are we using for determining validity? Eg:
```carbon
fn What(bound p: i32*) -> i32* {
  var x: i32 = 0;
  var q: i32* = &x;
  if (true) { q = p; }
  // OK?
  return q;
}
```

Author: I'm fine with whatever rule you want since these rules are going to be replaced by the memory safety story. I'm going to say it is erroneous behavior to violate the rules without saying anything about enforcement.

}

fn Invalid2(bound x: i32) -> ref i32 {
  var y: i32 = x;
  // ❌ Invalid.
  return y;
}

// ✅ Valid

> fn Valid1(bound p: i32*) -> ref i32 {

Reviewer 5: What does it mean to apply `bound` to a pointer, especially a pointer to a pointer? What does `bound ref p: i32*` mean?

If we follow the `clang::lifetimebound` model, `bound` would apply to the pointee of the top-level pointer only, but that's inconsistent with what it would mean for a non-copy-value-representation by-value function parameter, where it would apply to the lifetime of the pointer itself. Is it OK that we have that inconsistency (especially in generic code)?

Author: My reading of the `clang::lifetimebound` model, and what I've tried to propose is that `bound` means "return may reference bound parameters, or anything transitively reachable from them."

https://clang.llvm.org/docs/AttributeReference.html#lifetimebound

> By default, a reference is considered to refer to its referenced object, a pointer is considered to refer to its pointee, a std::initializer_list<T> is considered to refer to its underlying array, and aggregates (arrays and simple structs) are considered to refer to all objects that their transitive subobjects refer to.

Reviewer 5: `clang::lifetimebound` doesn't give transitive reachability, only direct reachability through that value (and through its subobjects if it's an aggregate). So for example:

* `[[clang::lifetimebound]] int **p` means that `*p` is bound and `**p` is not. (And also `p` itself is not, because that's not meaningful in C++.)
* `[[clang::lifetimebound]] X x`, with `struct X { int a; int *p; int **q; }` means that `*x.p` is bound, `*x.q`, but `**x.q` is not.
* Reference parameters are treated like pointer parameters.

(where "object X is bound" means "object X is required to live at least as long as the return value of the function"). `[[clang::lifetimebound]]` doesn't have a mechanism to constrain indirect lifetimes (eg, `int**` or `string_view`).

This doesn't quite match what we're doing in Carbon, because values also potentially have lifetime in Carbon (whereas in C++ function parameters are always `var`s), so for a `bound p: i32**`, we need to know whether `p` is bound, `*p` is bound, and/or `**p` is bound.

Reviewer 5: I'd be fine with saying that the exact semantics of `bound` are not being decided in this proposal and being left to a future memory safety proposal.

Reviewer 2: I've suggested an edit to capture this below.

Reviewer 5: Suggested edit works for me.

fn ValToRef(bound y: C) -> ref D { return *y.ptr; }
fn VarToRef(bound var p: i32*) -> ref i32 { return *p; }
fn VarToDefault(bound var p: i32*) -> i32* { return p; }
```

For full safety, we need each bound parameter to be immutable for the duration
of the lifetime of the returned result. However, the objective for now is only
matching `[[clang::lifetimebound]]`, which has the goal of preventing some
classes of bugs, not full memory safety. We will reconsider this with the memory
safety design.

### How addresses interact with `ref`

The address of a `ref` binding is `noalias` and either `captures(none)` or
`captures(ret: address, provenance)`, depending on whether the binding is marked
`bound`.

-   `noalias` means like C `restrict`; you can't observe mutations through
    aliases; mutation through a restricted pointer is not observable through
    another pointer
-   `captures(none)` means there is no transitive escape: you can pass a
    nocapture pointer to another nocapture function, but you can't store to
    memory or return
-   `captures(ret: address, provenance)`: is like `captures(none)` but may be
    referenced by a return.

> The `noalias` semantics are the minimum for the "move-in-move-out" optimization.

Reviewer 5: I think you also need `captures(none)`, otherwise the optimization is observable through address comparisons.

Author: Done.

for those cases.

### Part of the expression type system, not object types

Much like value/`val` and `var` bindings, `ref` binding and the new return forms
are are part of the type system, but only through expression categories,
patterns (function parameters and so on), and returns. Specifically, we don't
expect them to be part of the _object types_ in Carbon. Like value bindings, we
retain a great deal of implementation flexibility around layout, and the
specifics of how they are lowered.

This specifically means we will need to incorporate `ref` bindings into the
`Call` interface and we will be adding complexity there that will need to be
handled by overloading. The changes to the `Call` interface is future work, and
overloading, once we add support, will need to carry additional complexity to
handle `ref`.

### Interaction with `returned var`

The rule is: `returned var` may only be used when there is a single component to
the return form, and it is default `->`.

```carbon
fn F(...) -> V {
  returned var v: V = ...;
  // ...
  return var;
}

fn F(...) -> {-> .a : T} {
  returned var ret: T = ...;
  // ...
  return var;

> }

Reviewer 5: Is this second example OK? I'm not sure whether this constitutes "default `->`" or not. Also, is this using old syntax (`{-> ...}`)? Should that be `fn F(...) -> {var .a: T}`?

FWIW I'd prefer to reject this example. I find it strange that our `return var` has a different type than the return type of the function and that the function body never mentions that it's initializing a field named `a` at all, and given how narrow a case we can support here, it doesn't seem worth allowing this.

Author: Done.

```

We can revisit and expand this later if this does not handle use cases we would
like to support.

### Use case: `Deref` interface

To support customization of the prefix-`*` dereferencing operator, we introduce
the `Deref` interface.

```carbon
interface Deref {
  let Result:! type;
  fn Op[bound ref self: Self]() -> ref Result;
}

final impl forall [T:! type] T* as Deref {
  where Result = T;
  fn Op[bound ref self: Self]() -> ref T
      = "builtin.deref";
}
```

Then `*p` is rewritten to `p.(Deref.Op)()`, and `p->m` is rewritten to
`p.(Deref.Op)().m`. For example, this might be used by a smart pointer:

```
class SmartPtr(T:! type) {
  fn Make(p: T*) -> Self { return {.ptr = p}; }
  impl as Deref {
    where Result = T;
    fn Op[bound ref self: Self]() -> ref Result {
      return *self.ptr;
    }
  }
  private var ptr: T*;
}
```

> ### Use case: indexing interfaces

Reviewer 5: (Note to self: this is where I got to.)

  *y_ptr += 1;
}
```

We add support for `ref` and `var` in a
[struct pattern](/docs/design/pattern_matching.md#struct-patterns) when using
the shorthand `a: T` syntax for `.a = a: T`:

```carbon
let {var a: i32, ref b: i32} = ...;

// Now equivalent to:
let {.a = var a: i32, .b = ref b: i32} = ...;
```

> Note: This takes us one step closer to `{` ambiguity. Previously we could
> distinguish between a struct literal/pattern and a non-empty block with only
> up to two tokens of lookahead (the struct cases start with `.` or `_` or
> identifier followed by `:`, and the block cases don't). Now we have things
> like:
>
> ```carbon
> fn F() -> X { var a: i32 = 0; }
> ```
>
> ... where we're getting incrementally closer to ambiguity. We've got a few
> more steps before we get there, though, since we don't have an `X{...}`
> expression yet, and `var ...` is only allowed in struct patterns rather than
> struct expressions. So we're still fine, but this is cutting down our options
> for future syntactic expansion a little.

The `ref` modifier is forbidden on the bindings in `class` or struct type
fields.

```
var outer_size: i32 = 123;

class Invalid {
  // ❌ Invalid.

>   let ref invalid_ref_field: i32 = outer_size;

Reviewer 2: Maybe just to clarify:

```suggestion
  // ❌ Invalid. We don't currently have runtime `let` bindings in classes,
  // or `ref` on `var`s, but the intent is to not have `ref` bindings as fields.
  let ref invalid_ref_field: i32 = outer_size;
```

      -> ref Self;
}

impl C1 as I {
  // ✅ Valid: matches interface
  fn F[bound ref self: Self]
      (a: Self, bound b: Self, bound c: Self*)
      -> ref Self;
}

impl C2 as I {
  // ✅ Valid: proper subset of `bound` params
  fn F[ref self: Self]
      (a: Self, bound b: Self, c: Self*)
      -> ref Self;
}

impl C2 as I {
  // ❌ Invalid: `a` is not bound in `I.F`.
  fn F[ref self: Self]
      (bound a: Self, b: Self, c: Self*)
      -> ref Self;
}
```

Like `[[clang::lifetimebound]]` in C++, `bound` does not affect semantics or
calling conventions, just what code is legal. This helps avoid mismatches
between typechecking against the signatures in an interface when the `impl`
functions are different. Exception: the question of whether `bound` affects the
lifetime of temporaries is [future work](#temporary-lifetimes).

Note that all combinations of a `val`/`ref`/default return can be bound to a
value/`ref`/`var` parameter. Examples:

```carbon
fn RefToVal(bound ref x: C) -> val D { return x.d; }
fn ValToRef(bound y: C) -> ref D { return *y.ptr; }
fn VarToRef(bound var p: i32*) -> ref i32 { return *p; }
fn VarToDefault(bound var p: i32*) -> i32* { return p; }
```

> For full safety, we need each bound parameter to be immutable for the duration
> of the lifetime of the returned result. However, the objective for now is only
> matching `[[clang::lifetimebound]]`, which has the goal of preventing some
> classes of bugs, not full memory safety. We will reconsider this with the memory
> safety design.

Reviewer 2: Suggested edit to document the subtlety of model here:

```suggestion
For full safety, we need each bound parameter to be immutable for the duration
of the lifetime of the returned result. However, the objective for now is only
matching `[[clang::lifetimebound]]`, which has the goal of preventing some
classes of bugs, not full memory safety. We will reconsider this with the memory
safety design.

Clang's `lifetimebound` attribute also only applies to the immediately pointed to objects (by pointers or reference parameters, or pointers or reference subobjects of an aggregate parameter). We suggest a simpler, transitive model here that is more restrictive but should be compatible. That said, pinning down the exact and firm semantics of `bound`, especially in these complex cases, is deferred to the full memory safety design as well.
```

// ReturnLet() is a value expression.
let l: i32 = ReturnLet();

fn ReturnDefault() -> i32 {
  return 2;
}
// ReturnDefault() is an initializing expression.
var j: i32 = ReturnDefault();
```

-   A call to a function declared `-> ref T` is a durable reference expression.
    The generated code for that function will return the address of a `T`
    object.
-   A call to a function declared `-> let T` is a value expression. The function
    will return the value representation of `T`. Since values have no address,
    the value representation may be returned in registers.
-   The behavior of a call to a function declared `-> T` is unchanged. It is an
    initializing expression, returning in place or by copy depending on the
    initializing representation of `T`.
-   The behavior of `auto` as the return type is unchanged, but now supports an
    optional `ref` or `let` between the `->` and `auto`. `-> auto` continues to
    return an initializing expression, `-> let auto` returns a value expression,
    and `-> ref auto` returns a durable reference expression.

> -   Using `=>` to specify a return continues to return an initializing

Reviewer: I think there's a strong case to be made for having `=>` deduce the expression form, because that makes`(fn => expr)()` a drop-in replacement for `expr`. I'd be fine with leaving this as an open question, but I'd rather not resolve it (particularly not in this way) without considering that alternative.

Author: There is an issue that if we return something other than an initializing expression, we probably need the parameters to be marked `bound`.  I added some text saying that.
